open_lm/file_utils.py

```diff
@@ -390,6 +399,15 @@ def _single_epoch_string(
     needed_samples_per_source = [int(np.ceil(weights[i] * num_samples / sum(weights))) for i in range(num_sources)]
 
     manifests = [get_metadata_file(path) for path in paths]
+    if shard_shuffle_seed is not None:
+        # Not the most efficient but probably the safest since shuffling is done inplace
+        rng_gen = np.random.default_rng(shard_shuffle_seed)
+        new_manifests = []
+        for m in manifests:
```

Comment:
i think this will break if manifests are provided in a different order in the commandline? e.g., someone passes manifest A, then manifest B vs. the other way?

Reply:
to handle this potentially re-initialize `rng_gen` within the loop?

Reply:
also do we have to call shuffle more than once (here and in `get_shards_for_chunk`)? seems safer if we can just shuffle the source shards once and then "read off" the new order of shards everywhere else?

Reply:
Initializing rng_gen within the loop is correct, thanks for catching that!

I thought of doing the shuffling inside `get_metadata_file` (immediately after reading) but decided against it originally to preserve backwards compatibility with previous unshuffled runs. Now that I think about it again, passing it inside `get_metadata_file` is safer, as you said.